### PR TITLE
refactor: Serialization/deserialization for collection, thresholds, scatterplot, and backdrops (state pt. 13)

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Problem
=======
Part 13 of ??? for #492, "refactor state management."

This change adds serialization for the remaining state slices. This means that all viewer settings are now being pulled from the URL and set in the state on initial load.

The next PR will focus on the inverse relationship, syncing state to the URL.

*Estimated review size: medium, 30-40 minutes* (it's mostly unit tests)

Solution
========
- Added serialization and deserialization for the collection, threshold, scatterplot, and backdrop slices.
- Added unit tests.
- Fix for very small values in `numberToStringDecimal`.
- Refactored unit tests in `url_utils.ts` for threshold serialization/deserialization.

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Open the preview link. You should see the same settings in the scatter plot, filters, and viewer settings as the screenshots below:  https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-594/viewer?collection=https%3A%2F%2Fallencell.s3.amazonaws.com%2Faics%2Fnuc-morph-dataset%2Ftimelapse_feature_explorer_datasets%2Fexploratory_dataset%2Fcollection.json&dataset=Small&feature=volume&t=15&filters=height%3A%25CE%25BCm%3A7.840%3A14.733%2Cvolume%3A%25CE%25BCm%25C2%25B3%3A43.818%3A977.818%2Cfamily_id%3A%3A164%3A273&range=43.818%2C977.818&color=matplotlib-cool&palette-key=adobe&bg-sat=49&bg-brightness=132&fg-alpha=43&outlier-color=c0c0c0&outlier-mode=1&filter-color=dddddd&filter-mode=1&outline-color=ff00ff&tab=settings&vc=0&vc-color=000000&vc-key=_motion_&vc-scale=4&vc-time-int=5&vc-tooltip=m&bg=1&scalebar=1&timestamp=1&path=1&keep-range=0&bg-key=max_intensity_zprojection_of_lamin_b1&scatter-range=frame&scatter-x=change_in_volume_in_25_minute_window&scatter-y=height

**Backdrop:**
image

**Filters:**
image

**Scatter plot:**
image

### CODE REVIEW

File: src/colorizer/utils/url_utils.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ export enum UrlParam {
   BACKDROP_KEY = "bg-key",
   BACKDROP_BRIGHTNESS = "bg-brightness",
   BACKDROP_SATURATION = "bg-sat",
-  FOREGROUND_ALPHA = "fg-alpha",
```
Comment (ShrimpCryptid) [inline src/colorizer/utils/url_utils.ts:48]: Renamed the enum to match how we refer to this property everywhere else.

File: tests/url_utils.test.ts
Comment (ShrimpCryptid) [inline tests/url_utils.test.ts:1]: Now that `serializeThresholds` and `deserializeThresholds` are exported, I refactored the unit tests so they now test those methods directly instead of going through the (soon to be deprecated) `paramsToURLQueryString` method.

File: tests/state/ViewerState/threshold_slice.test.ts [hunk 2]
```diff
@@ -99,4 +113,29 @@ describe("ThresholdSlice", () => {
...
+      const { result } = renderHook(() => useViewerStateStore());
+      const params = new URLSearchParams();
+      params.set(UrlParam.THRESHOLDS, DEFAULT_THRESHOLDS_SERIALIZED);
+      act(() => {
+        result.current.setThresholds([]);
+        loadThresholdSliceFromParams(result.current, params);
+      });
+      expect(result.current.thresholds).to.deep.equal(DEFAULT_THRESHOLDS);
+    });
+  });
```
Comment (ShrimpCryptid) [inline tests/state/ViewerState/threshold_slice.test.ts:140]: Unit tests here are relatively minimal since I test most of the edge cases in `url_utils.test.ts`

File: tests/state/ViewerState/threshold_slice.test.ts
Comment (ShrimpCryptid) [inline tests/state/ViewerState/threshold_slice.test.ts:1]: Now that `serializeThresholds` and `deserializeThresholds` are exported, I refactored the unit tests so they now test those methods directly instead of going through the (soon to be deprecated) `paramsToURLQueryString` method.

File: tests/url_utils.test.ts
Comment (ShrimpCryptid) [inline tests/url_utils.test.ts:1]: Oops my comment didn't get saved. I refactored these tests now that `serializeThreshold` and `deserializeThreshold` are exported functions, so the tests no longer rely on the (soon to be deprecated) `paramsToURLQueryString` method.

File: src/colorizer/utils/math_utils.ts [hunk 3]
```diff
@@ -3,21 +3,41 @@ import { Vector2 } from "three";
...
 ): string {
   if (input === undefined || input === null) {
     return "NaN";
-  }
-  if (Number.isInteger(input) && skipIntegers) {
+  } else if (Number.isInteger(input) && skipIntegers) {
     return input.toString();
+  } else if (Math.abs(input) < 1) {
+    // For numbers less than 1, return value by precision
+    return input.toPrecision(maxSignificantDigitsAfterDecimal);
```
Comment (ShrimpCryptid) [inline src/colorizer/utils/math_utils.ts:37]: Fixed a bug here where very small numbers (like `0.000001`) would be truncated to `0.000`. It's possible that a feature might have a very very small range of values and the color ramp range + threshold would have been incorrect.

**Review by interim17 — APPROVED** [review]
Looks good and pretty similar to the last one, haven't run anything.

File: src/state/utils/store_io.ts [hunk 4]
```diff
@@ -38,19 +49,23 @@ export const serializedStoreDataToUrl = (data: SerializedStoreData): string => {
...
-  loadDatasetSliceFromParams(store.getState(), params);
   loadConfigSliceFromParams(store.getState(), params);
 
-  // 2. Dependent on dataset fields:
-  loadTimeSliceFromParams(store.getState(), params);
+  // 2. Dependent on dataset object:
+  loadBackdropSliceFromParams(store.getState(), params);
+  loadDatasetSliceFromParams(store.getState(), params);
+  loadScatterPlotSliceFromParams(store.getState(), params);
+  loadThresholdSliceFromParams(store.getState(), params);
```
Comment (toloudis) [inline src/state/utils/store_io.ts:63]: nit: would these be more specific if they passed in their relevant slices instead of the whole store?

**Review by toloudis — APPROVED** [review]

File: src/state/utils/store_io.ts (hunk 4 above)
Comment (ShrimpCryptid) [reply src/state/utils/store_io.ts:63]: Each of the slice loaders is generally parameterized by only that slice, with some exceptions for slices that need access to the dataset or other information. I wouldn't really be able to get only the individual slice contents unless I wrote selectors for each, which would be a big headache. 

The slices really only exist to separate out logic for readability, but they're all sort of mixed together into a big state soup in the actual store.
